sdk/appcenter/src/main/java/com/microsoft/appcenter/channel/DefaultChannel.java

```diff
@@ -454,50 +443,78 @@ private synchronized void triggerIngestion(final @NonNull GroupState groupState)
+                    authTokenContext.removeToken(authToken);
+                }
+                continue;
             }
-        }
+            AppCenterLog.debug(LOG_TAG, "ingestLogs(" + groupState.mName + "," + batchId + ") pendingLogCount=" + groupState.mPendingLogCount);
 
-        /* Remember this batch. */
-        groupState.mSendingBatches.put(batchId, batch);
+            /* Call group listener before sending logs to ingestion service. */
+            if (groupState.mListener != null) {
+                for (Log log : batch) {
+                    groupState.mListener.onBeforeSending(log);
+                }
+            }
 
-        /*
-         * Due to bug on old Android versions (verified on 4.0.4),
-         * if we start an async task from here, i.e. the async handler thread,
-         * we end up with AsyncTask configured with the wrong Handler to use for onPostExecute
-         * instead of using main thread as advertised in Javadoc (and its a static field there).
-         *
-         * Our SDK guards against an application that would make a first async task in non UI
-         * thread before SDK is initialized, but we should also avoid corrupting AsyncTask
-         * with our wrong handler to avoid creating bugs in the application code since we are
-         * a library.
-         *
-         * So make sure we execute the async task from UI thread to avoid any issue.
-         */
-        HandlerUtils.runOnUiThread(new Runnable() {
+            /* Remember this batch. */
+            groupState.mSendingBatches.put(batchId, batch);
+
+            /*
+             * Due to bug on old Android versions (verified on 4.0.4),
+             * if we start an async task from here, i.e. the async handler thread,
+             * we end up with AsyncTask configured with the wrong Handler to use for onPostExecute
+             * instead of using main thread as advertised in Javadoc (and its a static field there).
+             *
+             * Our SDK guards against an application that would make a first async task in non UI
+             * thread before SDK is initialized, but we should also avoid corrupting AsyncTask
+             * with our wrong handler to avoid creating bugs in the application code since we are
+             * a library.
+             *
+             * So make sure we execute the async task from UI thread to avoid any issue.
+             */
+            HandlerUtils.runOnUiThread(new Runnable() {
 
-            @Override
-            public void run() {
-                sendLogs(groupState, stateSnapshot, batch, batchId);
-            }
-        });
+                @Override
+                public void run() {
+                    sendLogs(groupState, stateSnapshot, batch, batchId, authToken);
+                }
+            });
+            return;
+        }
+
+        /* Nothing to send. Some corrupted entries may be deleted, reset the counter. */
+        groupState.mPendingLogCount = 0;
```

Comment:
That won't work with multiple batches in parallel. The code before was removing -= fetchSize, not just setting to 0. What is this change for?

Reply:
It will. It still decrements the size - see a line 482.
As discussed in slack - this logic is affected by time-filtering.